### PR TITLE
*: remove opt package totally

### LINKED ISSUE #4340: Remove opt package entirely
Previously, package `opt` is used to temporarily solve the circular dependency problem. It's not well abstracted and meanless, we should remove it.

- [x] https://github.com/tikv/pd/pull/4196
- [x] https://github.com/tikv/pd/pull/4517
- [x] https://github.com/tikv/pd/pull/4520
- [x] https://github.com/tikv/pd/pull/4527
- [ ] https://github.com/tikv/pd/pull/4535

### PR DESCRIPTION
**Pull request by rleungx**
### What problem does this PR solve?

Close #4340.

### What is changed and how it works?

This PR moves the `Cluster` interface into the `schedule` package and creates a new subinterface `ClusterInfo` to get necessary cluster information in `operator` package. Also, this PR remove some locks which could be unnecessary. It needs to be reviewed carefully.

### Check List

Tests

- Unit test

### Release note

```release-note
None
```

### CODE REVIEW

File: server/schedule/operator/step.go [hunk 1]
```diff
@@ -36,7 +35,7 @@ type OpStep interface {
 	fmt.Stringer
 	ConfVerChanged(region *core.RegionInfo) uint64
 	IsFinish(region *core.RegionInfo) bool
-	CheckInProgress(cluster opt.Cluster, region *core.RegionInfo) error
+	CheckInProgress(checkFn func(id uint64) error, region *core.RegionInfo) error
```
Comment (disksing) [inline server/schedule/operator/step.go:38]: `checkFn` is not easy to understand. maybe define a `type CheckStoreStateFn func(id uint64) error`.

**Review by disksing — APPROVED** [review]

File: server/cluster/cluster.go [hunk 2]
```diff
@@ -490,11 +483,30 @@ func (c *RaftCluster) SetStorage(s *core.Storage) {
 	c.storage = s
 }
 
+// GetBasicCluster returns the basic cluster.
+// There is no need a lock since it won't changed.
```
Comment (HunDunDM) [inline server/cluster/cluster.go:487]: Written in `RaftCluster.InitCluster`

File: server/cluster/cluster.go [hunk 3]
```diff
@@ -490,11 +483,30 @@ func (c *RaftCluster) SetStorage(s *core.Storage) {
...
+func (c *RaftCluster) GetAllocator() id.Allocator {
+	return c.id
+}
+
+// GetRuleManager returns the rule manager reference.
+// There is no need a lock since it won't changed.
+func (c *RaftCluster) GetRuleManager() *placement.RuleManager {
+	return c.ruleManager
+}
+
```
Comment (HunDunDM) [inline server/cluster/cluster.go:509]: ditto

File: server/schedule/operator/builder.go [hunk 4]
```diff
@@ -21,13 +21,22 @@ import (
...
 	"github.com/tikv/pd/server/core"
+	"github.com/tikv/pd/server/id"
 	"github.com/tikv/pd/server/schedule/filter"
-	"github.com/tikv/pd/server/schedule/opt"
 	"github.com/tikv/pd/server/schedule/placement"
 	"github.com/tikv/pd/server/versioninfo"
 )
 
+// ClusterInfo provides the necessary information for building operator.
+type ClusterInfo interface {
```
Comment (HunDunDM) [inline server/schedule/operator/builder.go:33]: Any other naming ideas?

File: server/schedule/operator/builder_test.go [hunk 5]
```diff
@@ -118,7 +119,6 @@ func (s *testBuilderSuite) TestPrepareBuild(c *C) {
 	_, err := s.newBuilder().SetPeers(map[uint64]*metapb.Peer{4: {StoreId: 4, Role: metapb.PeerRole_Learner}}).prepareBuild()
 	c.Assert(err, NotNil)
 
-	// use joint consensus
```
Comment (HunDunDM) [inline server/schedule/operator/builder_test.go:121]: Unexpected change?

File: server/schedule/operator/builder_test.go [hunk 6]
```diff
@@ -141,6 +141,7 @@ func (s *testBuilderSuite) TestPrepareBuild(c *C) {
 	c.Assert(builder.currentLeaderStoreID, Equals, uint64(1))
 
 	// do not use joint consensus
+	s.cluster.SetClusterVersion(versioninfo.MinSupportedVersion(versioninfo.Version4_0))
```
Comment (HunDunDM) [inline server/schedule/operator/builder_test.go:144]: ditto

File: server/schedule/operator/create_operator_test.go [hunk 7]
```diff
@@ -159,6 +159,8 @@ func (s *testCreateOperatorSuite) TestCreateSplitRegionOperator(c *C) {
 }
 
 func (s *testCreateOperatorSuite) TestCreateMergeRegionOperator(c *C) {
+	// use joint consensus
+	s.cluster.SetClusterVersion(versioninfo.MinSupportedVersion(versioninfo.JointConsensus))
```
Comment (HunDunDM) [inline server/schedule/operator/create_operator_test.go:163]: Why would this be needed?

File: server/schedule/operator/step.go [hunk 8]
```diff
@@ -672,14 +673,3 @@ func (cpl ChangePeerV2Leave) CheckInProgress(cluster opt.Cluster, region *core.R
 
 // Influence calculates the store difference that current step makes.
 func (cpl ChangePeerV2Leave) Influence(opInfluence OpInfluence, region *core.RegionInfo) {}
-
-func validateStore(cluster opt.Cluster, id uint64) error {
```
Comment (HunDunDM) [inline server/schedule/operator/step.go:676]: Can't use `ClusterInfo` to replace `opt.Cluster`?

Comment (rleungx) [reply server/schedule/operator/step.go:676]: We can but I think it's unnecessary.

File: server/schedule/operator/builder.go (hunk 4 above)
Comment (rleungx) [reply server/schedule/operator/builder.go:33]: No, I cannot find a better naming. Any suggestion?

File: server/cluster/cluster.go (hunk 3 above)
Comment (rleungx) [reply server/cluster/cluster.go:509]: This function will not be called before the cluster is initialized.

File: server/cluster/cluster.go (hunk 2 above)
Comment (rleungx) [reply server/cluster/cluster.go:487]: ditto

Comment (rleungx) [reply server/cluster/cluster.go]: After discussing with @HunDunDM, I'm going to raise another PR to do it.

**Review by HunDunDM — APPROVED** [review]

**Pull request merged**